### PR TITLE
Add modifications for HW CNODE capability

### PR DESCRIPTION
**Pull request by getwildr**
## Description
Added more HW-CLI capabilities besides importing wallets | signing with payment & stake keys.
i.e. ability to work with node keys

## Where should the reviewer start?
Register a pool and/or rotate key. I still need to implement some kind of import node.

## Motivation and context
I've realized node keys could be handled by HW-CLI as well by visiting their repository.

## Which issue it fixes?
n/a

## How has this been tested?
I've created manually the cold.hwsfile. Then rotate/register, and delegated to it (from both CLI and HW wallets as co-owners).
There is a lot to do yet but that's a start and I'm looking forward to have your feedback first and advice as to how to move forward from what has been done up to now.

### CODE REVIEW

**Comment by Scitz0** [comment]
Hi, Being able to use pool keys on HW is an excellent addition so you have my approval to continue with this implementation :)

I haven't gone through all pieces of the code but saw that you so far have added the ability to import pool cold keys to create a pool from hardware. Maybe this should be made clearer in the menu selection. Perhaps instead of an import selection in the main pool menu, have the "New - create a new pool" option bring you to a submenu like for wallet->import->hw/mnemonic.  But instead Pool -> New -> CLI / HW. As this will actually create a new pool from CNTools perspective, based on cold key from HW. 

I also saw the hardcoded stuff etc and noticed that it's a WIP currently so that's fine, won't comment on that.

**Comment by getwildr** [comment]
Hi @Scitz0,

Glad to see you're OK with the idea.

Regarding this:
"Perhaps instead of an import selection in the main pool menu, have the "New - create a new pool" option bring you to a submenu like for wallet->import->hw/mnemonic"

I did copy paste the wallet way.
I wondered first why we choose Import Wallet over New HW Wallet, but then I've thought OK, indeed the keys already exist, we don't create them, we import existing keys.
So I decided to the same path for pools. Keys already exist in a way.
Do you agree more now or do you still think we should change the things here ?

As per the rest, I'll keep adding code and we'll discuss later how we can optimize.

**Comment by rdlrt** [comment]
Can we rebase this on setup-upds branch? It'll likely create some conflicts once that branch is merged

**Comment by getwildr** [comment]
> Can we rebase this on setup-upds branch? It'll likely create some conflicts once that branch is merged

Sure. Who can do that ? You or me ? What should I do exactly in such case ?

**Comment by rdlrt** [comment]
> 

The owner of PR or any of the collabs can do it, have done it now.
How : Click on Edit  PR -> check pull down for base branch (was previously alpha) -> change to target branch -> resolve conflicts if any

**Comment by getwildr** [comment]
OK thanks. I did pull/fetch before doing anything in order to be up to date, what should I have done ? Checkout latest branch before creating my own ?

**Comment by getwildr** [comment]
OK thanks. I did pull/fetch before doing anything in order to be up to date, what should I have done ? Checkout latest branch before creating my own ?

**Comment by rdlrt** [comment]
> OK thanks. I did pull/fetch before doing anything in order to be up to date, what should I have done ? Checkout latest branch before creating my own ?

What you did is fine 👍🏻 , this was just a peculiar case as the two branches have deviated and setup_upds contains breaking changes which is good to be adopted for future PRs

**Comment by getwildr** [comment]
@rdlrt alright.
So I've been reading your PR, I like this kind of guild-tools 2.0 :smiley: 
But you say in it scurrent state, some scripts might not work but thats expected for the time being.
Should I expect any impacts on my tests ?
Also now you did some updates on PR, what should I do on my side, locally, to be phased with these changes ?

**Comment by rdlrt** [comment]
While in this case, git pull should be fine - in general if you dont have local changes, best to use `git reset --hard origin/<branch-name>` to ensure your history is exactly same as upstream (note that reset --hard will definitely discard and ignore any local changes you make that isn't already pushed)

**Comment by getwildr** [comment]
Well I think I've reached my goal.
It is now possible to import/rotate/register/modify/retire a pool, w/ or w/o a HW co owner wallet.
I'll check again that I can do all these steps once my pool is actively retired in two days.
I've done some non reg testing, checked that I could do these steps with a CLI pool and it was alright.
Anyway I'll let your review all my modifications and make remarks. Thanks in advance,

**Comment by Scitz0** [comment]
Sorry for not getting back to you on this change, have been a bit overworked as of late and just haven't had the time 😞 
Hopefully I will have some time this week.

File: scripts/cnode-helper-scripts/cntools.sh
Comment (Scitz0) [inline scripts/cnode-helper-scripts/cntools.sh:1707]: We might want to limit this to Ledger for now. I couldn't get Trezor to work for pool cold keys. Could be a problem with my Trezor and VM env but I don't think so. 

```bash
cardano-hw-cli node key-gen --path 1853H/1815H/0H/0H --hw-signing-file /opt/cardano/guild/priv/pool/Trezor/cold.hwsfile --cold-verification-key-file /opt/cardano/guild/priv/pool/Trezor/cold.vkey --operational-certificate-issue-counter-file /opt/cardano/guild/priv/pool/Trezor/cold.counter
Error: Extended public key export cancelled by user
```

**Review by Scitz0 — CHANGES_REQUESTED** [review]
I finally found a bit of time to look over the changes. I added a couple of comments, mostly minor stuff. Due to some issues with my hardware device and VM env I wasn't able to test it fully. But looking at changes it seems fine.

**Comment by getwildr** [comment]
Hi @Scitz0 
I won't have my HW test device for the next 2 weeks.
So I'll get back to work after I get back home.

File: scripts/cnode-helper-scripts/cntools.sh
Comment (getwildr) [reply scripts/cnode-helper-scripts/cntools.sh:2216]: Will try to add this.

File: scripts/cnode-helper-scripts/cntools.sh
Comment (getwildr) [reply scripts/cnode-helper-scripts/cntools.sh:2233]: idem

File: scripts/cnode-helper-scripts/cntools.sh
Comment (getwildr) [reply scripts/cnode-helper-scripts/cntools.sh:2267]: idem

File: scripts/cnode-helper-scripts/cntools.sh
Comment (getwildr) [reply scripts/cnode-helper-scripts/cntools.sh:1783]: Humm I don't quite master this. Can I let you change/check this ?

File: scripts/cnode-helper-scripts/cntools.sh
Comment (getwildr) [reply scripts/cnode-helper-scripts/cntools.sh:1743]: I could raise a flag if one of this fails, and eventually use this flag to safe del the dir if needed ?

File: scripts/cnode-helper-scripts/cntools.sh
Comment (getwildr) [reply scripts/cnode-helper-scripts/cntools.sh:1740]: ACK

File: scripts/cnode-helper-scripts/cntools.sh
Comment (getwildr) [reply scripts/cnode-helper-scripts/cntools.sh:1707]: Have you succeeded in having some output with basic "device" command if it exists with TREZOR ?

**Comment by Scitz0** [comment]
> Hi @Scitz0 I won't have my HW test device for the next 2 weeks. So I'll get back to work after I get back home.

Thats fine, I made the changes according to comments. Slightly modified. Also decided to redo how cancel on wallet|pool|policy|asset selection is handled so ended up changing quite a few lines. Just need to test that I didn't break anything.

**Comment by getwildr** [comment]
Thanks a lot !

File: scripts/cnode-helper-scripts/cntools.sh
Comment (rdlrt) [inline scripts/cnode-helper-scripts/cntools.sh:1668]: Not sure if we should commit trezor not supported string completely, as it's a state that can change again in future with minimal/no change to code

File: scripts/cnode-helper-scripts/cntools.sh
Comment (rdlrt) [inline scripts/cnode-helper-scripts/cntools.sh:1679]: same

File: scripts/cnode-helper-scripts/cntools.sh
Comment (rdlrt) [inline scripts/cnode-helper-scripts/cntools.sh:657]: given the amount of time these 4 lines are repeating, makes sense to convert it to function instead?

File: scripts/cnode-helper-scripts/cntools.sh
Comment (Scitz0) [reply scripts/cnode-helper-scripts/cntools.sh:1668]: Pool keys on Ledger was added a long time ago. Over a year I think. I haven't heard of it being worked on for trezor so my guess is it wont be added any time soon, if ever. 

If we dont include any text about it users might think trezor should be supported.

File: scripts/cnode-helper-scripts/cntools.sh
Comment (Scitz0) [reply scripts/cnode-helper-scripts/cntools.sh:657]: Can you break out of loop in calling code with continue from within a function call? 🤔

I guess I could try.

File: scripts/cnode-helper-scripts/cntools.sh
Comment (rdlrt) [reply scripts/cnode-helper-scripts/cntools.sh:1668]: maybe can use as below , specifying what we can test - and leave the rest to upstream?
```suggestion
						" ) Import   - import node cold keys from HW device (only Ledger has been tested)"\
```

Comment (Scitz0) [reply scripts/cnode-helper-scripts/cntools.sh]: Though I understand your intention I want to leave text to make it clear that only Ledger is supported. But I rephrased it a bit. If Trezor should decide to support this down the road, its a small change to update text. Rephrasing it as only Ledger tested still suggests that Trezor could work while it's not supported on the device.

File: scripts/cnode-helper-scripts/cntools.sh
Comment (Scitz0) [reply scripts/cnode-helper-scripts/cntools.sh:657]: Though a bit spammy, there is no good way to reduce this code piece as you can't manipulate a loop from a function call. Because of this I will leave it as is for now and merge.

**Review by Scitz0 — APPROVED** [review]

**Pull request merged**
